File: backend-java/src/main/java/com/github/pascualex/oc/mapper/CourseMapper.java
package com.github.pascualex.oc.mapper;

import com.github.pascualex.oc.model.Course;
import org.apache.ibatis.annotations.Mapper;
import org.apache.ibatis.annotations.Param;

import java.util.List;

@Mapper
public interface CourseMapper {

    List<Course> getCourses();

    List<Course> getCoursesById(@Param("id") int id);

    void addCourse(@Param("course") Course course);

    void setCoursesActiveById(@Param("id") int id, @Param("active") boolean active);

    int removeCoursesById(@Param("id") int id);
}

File: backend-node/src/mapper/courseMapper.ts
import { database } from '../utils/Database';
import { Course, CourseRow } from '../model/Course';

export async function getCourses(): Promise<CourseRow[]> {
  return database.query(`
    SELECT c.id, c.title, c.level, c.hours, c.active, t.id as teacher_id, t.name as teacher_name
    FROM courses AS c LEFT JOIN teachers AS t ON t.id=c.teacher_id
    ORDER BY c.title, t.name, c.active, c.level, c.hours, c.id, t.id
  `);
}

export async function getCoursesById(id: number): Promise<CourseRow[]> {
  return database.query(`
    SELECT c.id, c.title, c.level, c.hours, c.active, t.id as teacher_id, t.name as teacher_name 
    FROM courses AS c LEFT JOIN teachers AS t ON t.id=c.teacher_id
    WHERE c.id=${id}
    ORDER BY c.title, t.name, c.active, c.level, c.hours, c.id, t.id
  `);
}

export async function addCourse(course: Course): Promise<void> {
  return database.query(`
    INSERT INTO courses (title, level, hours, active, teacher_id)
    VALUES ("${course.title}", "${course.level}", ${course.hours}, ${course.active}, ${course.teacher.id})
  `);
}

export async function setCoursesActiveById(id: number, active: boolean): Promise<void> {
  return database.query(`
    UPDATE courses
    SET active=${active}
    WHERE id=${id}
  `);
}

export async function removeCoursesById(id: number): Promise<void> {
  return database.query(`
    DELETE FROM courses
    WHERE id=${id}
  `);
}